### ISSUE TITLE
Sampling from priors is slow

### ISSUE THREAD

**Issue by tmalsburg** [paul-buerkner/brms#606]
I'm sampling from the priors via `sample_prior="only"` in order to perform prior predictive checks.  Perhaps I'm doing something wrong but unfortunately this takes a really long time, as long as fitting the model or perhaps even longer.  Shouldn't sampling from the priors be super fast?  The thing with prior predictive checks is that you typically do a lot of them when searching for appropriate priors.  However, given how slow this is, it's not really feasible in brms.  By the way, thanks for brms.  Amazing package!

**Comment by tmalsburg** [comment]
I also get warning messages reporting divergent transitions and transitions exceeding the maximum treedepth.  If I understand correctly, this shouldn't happen when we just sample the priors.

**Comment by paul-buerkner** [comment]
I would generally prefer questions being asked at https://discourse.mc-stan.org/. Would you mind asking the question there again and also provide information about the priors you are using?

**Comment by tmalsburg** [comment]
I thought of this as a bug report not a question :) but sure, will do.

**Comment by paul-buerkner** [comment]
Oh in that case, of course a github issue is ok. It's unlikely to be a bug though, but related to the kind of priors you are using.

**Comment by tmalsburg** [comment]
Okay, I'll post a more detailed description here but I'll be happy to move to discourse when you think that's a better place for this converation.

``` R
brm(reading.time ~ dependency * clause +
                  (dependency * clause | sid) +
                  (dependency * clause | item),
    data   = expt2_spr,
    family = lognormal,
    prior  = c(
      set_prior("normal(0, 1)",   class="b"),
      set_prior("normal(0, 0.5)", class="sd"),
      set_prior("lkj(8)",         class="cor")),
    warmup = 1000,
    iter   = 4000,
    sample_prior = "only") -> p.expt2_spr
```

- `reading.time` is a reaction time in ms
- `dependency is a two-level factor, coded -0.5 and 0.5`
- `clause is a three-level factor, coded with a treatment contrast
- crossed random effects:
  - `sid` are participants
  - `item` are experimental items

Warning messages were:
```
Warning messages:
1: There were 2 divergent transitions after warmup. Increasing adapt_delta above 0.8 may help. See
http://mc-stan.org/misc/warnings.html#divergent-transitions-after-warmup 
2: There were 886 transitions after warmup that exceeded the maximum treedepth. Increase max_treedepth above 10. See
http://mc-stan.org/misc/warnings.html#maximum-treedepth-exceeded 
3: Examine the pairs() plot to diagnose sampling problems
```

Three chains took ~10min and one ~100min.  When I actually fit the model all chains take about 20min each.

**Comment by paul-buerkner** [comment]
The problem is likely with a prior you did not specify yourself. In your case, the prior of the overall intercept, which is not part of the `b` class. Look at `prior_summary` or `get_prior` for information on the default prior.
I suggest you set a prior to the `Intercept` class, as well. I hope this should fix your convergence problems

**Comment by tmalsburg** [comment]
Thank you.  Just to make sure I understand this:  The issue is that the prior is vague and HMC struggles with vanishing gradients?  Something like this?  I guess I was surprised that we need HMC at all to sample from the prior.

**Comment by paul-buerkner** [comment]
If you use `sample_prior = "only"` we fit the same model as usual but just omitting the likelihood statement (see `stancode(<model>)` for how its done). The amount of work necessary to change all of this to run in the generated quantities block and make the model be exactly the same as before, just without the likelihood, is not doable for the generality of models supported by brms.

**Comment by tmalsburg** [comment]
Okay, I tried it with a prior for `Intercept` and also for `sigma` and it does speed things up a bit.  Also no more warnings about bad transitions.  That's good.  However, it's still painfully slow and I had to use informative priors to make it work.  I don't think this would work with any vague/diffuse/non-informative priors.  I get your point about how difficult it would be to implement this in some more efficient way, but it *is* a bit of a pity that we can't really use this to evaluate priors unless when we already have a good idea of what they should be (which is the boring case).  As far as I'm concerned we can close this issue.  Thank you for your input!
